smarts/env/intersection_env.py

```diff
@@ -0,0 +1,196 @@
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+# THE SOFTWARE.
+
+import pathlib
+from typing import Any, Dict, Optional, Tuple
+
+import gym
+
+from smarts.core.agent_interface import (
+    OGM,
+    RGB,
+    AgentInterface,
+    DoneCriteria,
+    DrivableAreaGridMap,
+    NeighborhoodVehicles,
+    Waypoints,
+)
+from smarts.core.controllers import ActionSpaceType
+from smarts.env.hiway_env import HiWayEnv
+from smarts.env.wrappers.format_obs import FormatObs
+from smarts.env.wrappers.single_agent import SingleAgent
+from smarts.sstudio import build_scenario
+from smarts.zoo.agent_spec import AgentSpec
+
+
+def intersection_env(
+    headless: bool = True,
+    visdom: bool = False,
+    sumo_headless: bool = True,
+    envision_record_data_replay_path: Optional[str] = None,
+    img_meters=64,
+    img_pixels=256,
+):
+    """An intersection environment where a single agent needs to make an
+    unprotected left turn in the presence of traffic and without traffic
+    lights. Traffic vehicles stop before entering the junction.
+
+    Observation:
+        A `smarts.env.wrappers.format_obs:StdObs` dict, containing enabled keys,
+        is returned as observation.
+
+    Actions:
+        Type: gym.spaces.Box(
+            low=-1.0, high=1.0, shape=(3,), dtype=np.float32
+        )
+
+        Action     Value range
+        Throttle   [ 0, 1]
+        Brake      [ 0, 1]
+        Steering   [-1, 1]
+
+    Reward:
+        Reward is distance travelled (in meters) in each step, including the
+        termination step.
+
+    Episode termination:
+        Episode is terminated if any of the following occurs.
+        + Steps per episode exceed 3000.
+        + Agent collides, drives off road, drives off route, drives on shoulder,
+          or drives on wrong way.
+
+    Solved requirement:
+        If agent successfully navigates the intersection then `info["score"]`
+        will equal 1, else it is 0. Considered solved when `info["score"] == 1`
+        is achieved over 800 consecutive episodes.
+
+    Args:
+        headless (bool, optional): If True, disables visualization in
+            Envision. Defaults to False.
+        visdom (bool, optional): If True, enables visualization of observed
+            RGB images in Visdom. Defaults to False.
+        sumo_headless (bool, optional): If True, disables visualization in
+            SUMO GUI. Defaults to True.
+        envision_record_data_replay_path (Optional[str], optional):
+            Envision's data replay output directory. Defaults to None.
+
+    Returns:
+        A single-agent unprotected left turn intersection environment.
+    """
+
+    scenario = [
+        str(
+            pathlib.Path(__file__).absolute().parents[2]
+            / "scenarios"
+            / "intersections"
+            / "1lane_left_turn"
+        )
+    ]
+    build_scenario(scenario)
```

Comment:
My concern here is that this scenario does not exist in the package version. We could add a `smarts.scenario` module for it.

Reply:
In order to enable users to easily instantiate an environment via
```
env = gym.make("smarts.env:intersection-v0")
```
we would need to automatically include and build the scenario. 

To do so, we may
1) include the scenario into smarts package
    + This pull request currently includes the relevant scenario into `smarts` package.
    https://github.com/huawei-noah/SMARTS/blob/41af666cbd26c9c0eb098c7d1f89ccc858b07913/MANIFEST.in#L6
2) include a `smarts.scenario` module
    + I believe we attempted this method previously, but it required folder structure changes and more `setup.py` files.
3) `curl` install the scenario from another github repository during the call to `gym.make()`
    + Alternatively, we may try this method.

Reply:
1) I think might not be viable, including a directory called `scenarios` has the possibility to interact badly with other modules.
2) If possible we can include `smarts.scenarios` such like `recursive-include smarts/scenarios/intersections/1lane_left_turn *.xml *.py`. In this case it would not require a new `setup.py`.
3) This seems reasonable, although it perhaps breaks packaging standards so it is only useful if we make it a resource standard within SMARTS.

Reply:
I am not comfortable with merging this until the scenario issue is resolved.

Reply:
I have repackaged intersection-v0 scenario to be located inside `smarts.scenarios` package.